src/jit/hwintrinsiccodegenxarch.cpp

```diff
@@ -58,6 +58,11 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
     int                 ival        = Compiler::ivalOfHWIntrinsic(intrinsicID);
     int                 numArgs     = Compiler::numArgsOfHWIntrinsic(intrinsicID);
 
+    if (numArgs == -1)
+    {
+        numArgs = Compiler::numArgsOfHWIntrinsic(intrinsicID, node->gtSIMDBaseType);
```

Comment:
I'm not fond of the -1 return value as an indicator that you need to call the overload with the base type. And I think it's debatable whether it's any more efficient to have the extra parameter in the method (handling the case there) versus having the branch on the return value in the calling method, and I think the former is clearer.

Reply:
> I'm not fond of the -1 return value as an indicator that you need to call the overload with the base type.

@CarolEidt 

-1 return value was introduced earlier to indicate type dependent number of arguments see comment: 
https://github.com/dotnet/coreclr/blob/ca76af95c584797ede0930034c6892a6c252174d/src/jit/hwintrinsiclistxarch.h#L20

There are other intrinsics using -1 in Sse, Sse41 and implementation of `SetVector128` / `SetVector256` in Sse2 and Avx2 will require this as well.

> I think it's debatable whether it's any more efficient to have the extra parameter in the method (handling the case there) versus having the branch on the return value in the calling method, , and I think the former is clearer.

I understand that you mean `latter` not `former`. Assuming `latter` IMO we could handle it more cleanly with method call due to the increasing number of intrinsics and cases which would need special handling.

Please let me know your preference for solution here

Reply:
I meant the *former* i.e have a single method, possibly witg default parameter, and do the check and branch in the *called* method instead of the *caller*

Reply:
I meant the *former* i.e have a single method, possibly witg default parameter, and do the check and branch in the *called* method instead of the *caller*

Reply:
OK got it